Include the product factory and exporter for product without variant

thisiscab opened the pull request:
This will include code that will let us import products that have no
variants and that has variants. 

The next step will be to link the creation / save of a new
product with an event that will queue the exportation to Shopify.

https://app.clubhouse.io/dynamomtl/story/33/as-a-admin-when-i-create-a-product-with-a-single-sku-it-gets-created-in-shopify-so-that-i-don-t-need-to-create-it-twice

https://app.clubhouse.io/dynamomtl/story/35/as-a-admin-when-i-create-a-product-with-variants-the-master-product-and-its-variant-skus-are-created-in-shopify-so-that-i

thisiscab committed:
Include the product factory and exporter for product without variant

This will include code that will let us import products that have no
variants. The next step will be to link the creation / save of a new
product with an event that will queue the exportation to Shopify.

https://app.clubhouse.io/dynamomtl/story/33/as-a-admin-when-i-create-a-product-with-a-single-sku-it-gets-created-in-shopify-so-that-i-don-t-need-to-create-it-twice

thisiscab committed:
Export the product to Shopify when product is saved

And modify related specs to support that

thisiscab committed:
Add requests specs related to the product exportation

I wanted to make sure that we were correctly updating the product and
not creating new one when that was the case.

I have also decided to return the shopify_order instead of a true/false
response that I couldn't use.

thisiscab committed:
Add variant factory with related specs

That way we will be able to import the variant associated with the
product.

The only problem I have right now is related to the images. I'll
investigate later.

https://app.clubhouse.io/dynamomtl/story/35/as-a-admin-when-i-create-a-product-with-variants-the-master-product-and-its-variant-skus-are-created-in-shopify-so-that-i

thisiscab committed:
Modify the product factory to include the variant

thisiscab committed:
Add request specs for product with variants

Modify code to make those specs work

thisiscab committed:
Add specs related to updating an existing variant

We want to make sure that it doesn't duplicate or re-create the variant
when it is already existing on a product.

thisiscab committed:
WIP - first draft for the image import

There are tons of problems coming up. Mainly, we have a weird way of
setting and fetching the images on Glossier, I am still unsure how we'll
go around that in order to find the proper images. Right now I am using
placeholder images.

Also, in order to assign an image to a variant, the variant has to be
already saved. So we need to save the product, that saves the variant
then we assign the images to the product and when we save the product
again, it will assign the image to the variant. It is a really weird
process but I haven't found a way around that... Yet!

thisiscab committed:
Refactor the way that product factory image works

It creates a single ShopifyImage instead of an array of ShopifyImages.

I have also written specs to ensure that everything was working properly

thisiscab committed:
Save the request specs in cassettes

We don't really care about hitting the Shopify API every single time we
run those requests specs. It was taking over 40 seconds to run those
tests and now it runs almost instantly by using VCR.

Why VCR? Because this is what I am used to use and it accomplishes what
I want to do quickly.

thisiscab committed:
Use open instead of File.open

That way we can open the online URLs instead of just local files.

thisiscab committed:
Ensure that that the master variant is exported

I had totally forgotten to export the master variant of a product.
Oopsies.

thisiscab committed:
Make sure to re-create product/variant if not exist anymore

If we were to create a product on Solidus and sync it with Shopify then
delete the product from Shopify, the next time we would try to sync the
product/variant it would just crash.

Shopify would return an exception and would halt the process, in that
situation we catch the error and we re-create it to make sure that the
product exists in Shopify.

thisiscab committed:
Add more specs to make sure that everything works

Decided to temporarily remove VCR because it caused a lot of problems.
Since factory girl creates a master variant with a random sku and all
the other variants with random skus, it's hard to sync of with VCR.

I find it insane to mock each skus individually, there must be a better
way.

thisiscab committed:
Use the sku as the shopify option1 field if no option values

The option1 field in shopify is required. We are currently filling it
with option values of the solidus variant.

But the option values of a solidus variant are not required, thus we
need to use another field thus we are using the sku as the unique
identifier.

thisiscab committed:
Always use the SKU as the first option1

Shopify doesn't check for the combination of options to check if the
variant is unique, it only checks for the first option.

To ensure that it is unique per product, I have decided to always go
with the sku.

thisiscab commented:
Right now:
- VCR has been removed, please see the respective commit message
- Some tests were skipped because I have not taken the time to figure out how to fix them.

thisiscab committed:
Update travis file for specs to pass

The ENV variables were missing.

thisiscab committed:
Re-enable job the that export to shopify on product save

DanielWright commented on app/models/product_decorator.rb:
```diff
@@ -0,0 +1,9 @@
+Spree::Product.class_eval do
```
`Module#prepend` please! 😁

DanielWright commented on app/models/variant_decorator.rb:
```diff
@@ -0,0 +1,9 @@
+Spree::Variant.class_eval do
```
`Module#prepend` pleeeeeaaaase! 😁 😁

DanielWright commented on app/factories/shopify/variant_factory.rb:
```diff
@@ -0,0 +1,31 @@
+      shopify_variant.option1 = spree_variant.sku
+
+      spree_variant.option_values.each_with_index do |option, index|
+        shopify_variant.send("option#{index + 2}=", option.name)
```
What does this magic do?

thisiscab replied:
Where would you want that to be documented? In the code? It's already in the commit message.

DanielWright replied:
> In the code?

Actually, yes. One of the beauties of Ruby is how expressive it can be, and how that expressiveness can be used to write code that _documents itself_.

thisiscab replied:
I don't believe in code that documents itself, I'll write comments.

DanielWright replied:
I gotta tell you, this response blew my mind a bit.

Why would you double the maintenance workload by writing comments instead of self-documenting code? This introduces a risk of desynchronizing the code and the documentation. An additional benefit of writing expressive code is its expressiveness exists in two places – where the methods exist, and where the methods are invoked. The code can become almost human-readable, which is so much friendlier for future maintainers and integrators.

thisiscab replied:
Sorry for the comment I wrote, re-reading it and talking with Braiden made me realize that it was bad manner and it didn't leave place for discussion. I would have probably re-read it and made my point a little bit more explanatory. 

What I should have said is the following: 

I actually think that self-documenting code is kind of a myth. Something can't be clear for the majority of people. There will always be something that will make you have to dig anyway.

I just thought that we had that conversation many times in the past and we had decided to write comments when cases were nebulous and that's what I meant by writing "I'll write comments."

But at the same time, I remember having a conversation, I think in Frontendo, where we said that comments should be kept at a minimum.

Maybe I am just bad at finding write variables / method names to propagate the thinking behind the value of that said code.

But I would agree that what you have suggested below: `assign_variant_uniqueness_constraint`is speaking by itself and is far better than having a comment written. I jumped the gun too quickly before actually thinking about it.

My apologies for any confusions and I'll adjust for later comments! 💃

DanielWright commented on app/factories/shopify/variant_factory.rb:
```diff
@@ -0,0 +1,31 @@
+module Shopify
+  class VariantFactory
```
This doesn't seem like a Factory (which, conventionally, creates new objects) rather than a Converter or Migrator or Copier. It encapsulates the knowledge of how to mutate a Shopify variant into a Spree variant.

DanielWright commented on app/factories/shopify/variant_factory.rb:
```diff
@@ -0,0 +1,31 @@
+    attr_accessor :spree_variant, :shopify_variant
+
+    def generate_options!
+      shopify_variant.option1 = spree_variant.sku
```
Why does a Shopify variant need the SKU assigned to `option1` and `sku`?

thisiscab replied:
Where would you want that to be documented? In the code? It's already in the commit message.

DanielWright replied:
I take your point. But, imagine yourself as a new collaborator on the project in six months, or an end-user of the integration in a year. What would make you feel like this idiosyncrasy is well-documented?

thisiscab replied:
It was a genuine question, I usually have `NOTE (cab):` comments but it looks like we have tried to remove that kind of behavior in the code but I am unsure how else we can go about those stuff. 

For myself, I would have `NOTE(cab):` everywhere that stuff isn't totally clear like water!

DanielWright replied:
I think I'd try to hide this information in a separate method, maybe along the lines of:

```
def assign_variant_uniqueness_constraint(value)
  shopify_variant.option1 = value
end
```

I get several benefits from this:
1. The method is the sole owner of the knowledge that Shopify's uniqueness constraint is the value of `option1`.
2. It expresses the intent of the assignment. When `generate_options!` (or even `perform`) contains `assign_variant_uniqueness_constraint`, a first-time reader of the code has a better notion of what's happening.
3. I can expand this method with additional logic without making `generate_options!` more unwieldy. For example, I could perform some actual validation of the SKU's uniqueness.
4. It presents a more abstract interface, and makes the object a bit less Shopify-centric.

DanielWright commented on app/factories/shopify/product_image_factory.rb:
```diff
@@ -0,0 +1,38 @@
+      # When the image is hosted locally, we must get the `path` of the image,
+      # else if it's hosted online we must get the `url` of the image.
+      # I haven't found a better way of doing that yet.
+      begin
```
Clever! You might consider using `image.attachment.exists?` and `image.attachment.copy_to_local_file`, which seems to be an  interface method you can rely on in any of Paperclip's storage backends:
- [S3](https://github.com/thoughtbot/paperclip/blob/master/lib/paperclip/storage/s3.rb#L408)
- [Fog](https://github.com/thoughtbot/paperclip/blob/master/lib/paperclip/storage/fog.rb#L169)
- [Filesystem](https://github.com/thoughtbot/paperclip/blob/master/lib/paperclip/storage/filesystem.rb#L84-L86)

thisiscab replied:
That's interesting. `copy_to_local_file`

DanielWright commented on app/factories/shopify/product_factory.rb:
```diff
@@ -0,0 +1,59 @@
+    def find_or_initialize_variant(spree_variant)
+      begin
+        shopify_variant = ::ShopifyAPI::Variant.find(spree_variant.pos_variant_id) if spree_variant.pos_variant_id
+      rescue ActiveResource::ResourceNotFound
```
Barf. `ShopifyAPI::Variant.find` throws an exception? Is there a non-whiny version, like `.find_by` vs. `find_by!` in ActiveRecord?

thisiscab replied:
No sir :(

DanielWright replied:
Gross. Using exception-handling as control flow is such a bummer. 😞

DanielWright commented on app/factories/shopify/product_factory.rb:
```diff
@@ -0,0 +1,59 @@
+      shopify_product.created_at = spree_product.created_at
+      shopify_product.updated_at = spree_product.updated_at
+      shopify_product.published_at = spree_product.available_on
+      shopify_product.vendor = 'Glossier'
```
I think we should avoid hard-coding integration-specific values into this extension. Where could this value come from instead?

DanielWright commented on app/factories/shopify/product_factory.rb:
```diff
@@ -0,0 +1,59 @@
+    end
+
+    def surround_by_p_tags(content)
+      [content.lines.map { |line| "<p>#{line.strip}</p>" }].join
```
Pipe this through a Markdown parser, maybe?

DanielWright replied:
Better yet, what do you think of injecting a `parser` object that responds to `.parse`?

thisiscab committed:
Add a bunch of modifications based on CR

Mainly:
* Change factories to converters
* Better implementation for ShopifyVariant image uploads
* Better naming for certain methods
* Use prepend instead of class_eval for model decorators

thisiscab committed:
Allow the passing of a HTML renderer for the product converter

Instead of having our own method that makes it works only under a
certain scenario, we have opted to go with Redcarpet and convert the
Markdown content to HTML content. That way the content will be displayed
better than our previous ways of doing things.

Also, this let the user change the renderer to whatever rows their boat.

* This also moves the gems from the gemfile to the gemspec file.

thisiscab committed:
Add default value to the type of weight a variant has

thisiscab committed:
Fix all the specs for the product converter

There is a bunch of overhead and refactor that has to be done. There are
too many doubles and contexts everywhere.

thisiscab committed:
Fix the specs related to the variant converter

thisiscab committed:
WIP - Refactor the classes into more specialized classes

Still working on making it more readable and fixing the request specs
along the way.

thisiscab committed:
Refactor more stuff and fix more specs

thisiscab committed:
Fix the requests specs

All the requests specs are green, which means that we can continue
refactoring and create unit tests.

I have split up the requests specs in multiple different files like
suggested in the code-review.

thisiscab committed:
Change the behavior of the converter objects

This is a follow-up to what DW has suggested me

> There should already be a .update or .update_attributes method on `ShopifyAPI::Product`, as it extends ActiveResource
> https://github.com/rails/activeresource/blob/4-stable/lib/active_resource/base.rb#L1431-L1433

That way we can more easily update ShopifyAPI::Product. This breaks all
the requests specs but adds unit tests for each of those converters.

thisiscab committed:
Decorate the ActiveResource BaseDecorator

This will add the find_by and find_or_initialize_by method that will
help us further down the road.

thisiscab commented:
This will be split in multiple different PR
